packages/flutter/lib/src/widgets/scrollbar.dart

```diff
@@ -580,21 +616,69 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
   /// thumbOffsetLocal is a position in the thumb track. Cannot be null.
   double getTrackToScroll(double thumbOffsetLocal) {
     assert(thumbOffsetLocal != null);
-    final double scrollableExtent = _lastMetrics!.maxScrollExtent - _lastMetrics!.minScrollExtent;
+    late double scrollableExtent;
+    // If the ScrollView has a finite extent, and is not being overridden by
+    // infinite scroll, use the known extents.
+    if (_lastMetrics!.maxScrollExtent.isFinite && infiniteBehavior == null) {
+      scrollableExtent = _lastMetrics!.maxScrollExtent - _lastMetrics!.minScrollExtent;
+    } else {
+      // If we have an infinitely long scroll view (or are pretending to have
+      // one), we pretend it isn't by keeping track of the last known depth, and
+      // adding on more 'extent' based on the InfiniteScrollBehavior.
+      scrollableExtent = _infiniteDepth - _lastMetrics!.minScrollExtent;
+    }
     final double thumbMovableExtent = _trackExtent - _thumbExtent();
-
     return scrollableExtent * thumbOffsetLocal / thumbMovableExtent;
   }
 
   // Converts between a scroll position and the corresponding position in the
   // thumb track.
   double _getScrollToTrack(ScrollMetrics metrics, double thumbExtent) {
-    final double scrollableExtent = metrics.maxScrollExtent - metrics.minScrollExtent;
+    late double scrollableExtent;
+    final double trueScrollableExtent = metrics.maxScrollExtent - metrics.minScrollExtent;
+    // If the ScrollView has a finite extent, and is not being overridden by
+    // infinite scroll, use the known extents.
+    if (metrics.maxScrollExtent.isFinite && infiniteBehavior == null) {
+      scrollableExtent = trueScrollableExtent;
+    } else {
+      // If we have an infinitely long scroll view (or are pretending to have
+      // one), we pretend it isn't by keeping track of the last known depth, and
+      // adding on more 'extent' based on the InfiniteScrollBehavior.
+      // Update the greatest know depth.
+      switch (infiniteBehavior) {
+        case InfiniteScrollBehavior.page:
+          if (_infiniteDepth < metrics.pixels) {
+            _infiniteDepth = metrics.pixels;
```

Comment:
`_infiniteDepth` should be reset if the remote scrollable widget is removed or changed

Reply:
This is a really interesting point. The scrollbar does not currently check if the scrollable widget it is listening to changes. It listens to scroll events and updates accordingly, regardless of where (or which widget) those scrolling events are coming from. I wonder if there are other opportunities where we would want to reset like this.

Reply:
Maybe we can compare the `notificationContext`.